File: projections/react/ui/hlp.ui.app-shell/src/pack-navigation-mapper.ts
import { roleGateAllows, type HeldRoleSet, type RoleReference, type RoleVocabulary } from '@harborline-software/contracts/authorization'
import type { PackNavigationAction, PackNavigationDeclaration, PackPanelDeclaration, ShellNavigationState, ShellNavItem } from './types'

export interface ShellActionViewModel { id: string; label: string; icon: string; binding: string; shortcut: string }
export interface ShellGroupViewModel { id: string; label: string; items: readonly ShellNavItem[]; addAction?: ShellActionViewModel }
export interface ShellWorkspaceViewModel { id: string; label: string; icon?: string; count?: string | number; groups: readonly ShellGroupViewModel[]; createActions: readonly ShellActionViewModel[]; defaultCreateActionId?: string; createGuidance?: string; documentSpine: NonNullable<PackNavigationDeclaration['seedWorkspaces'][number]['documentSpine']>; recent: readonly ShellNavItem[]; suggested?: ShellNavItem }
export interface ShellNavigationViewModel { workspaces: readonly ShellWorkspaceViewModel[]; modes: readonly { id: string; label: string; workspaceIds: readonly string[] }[]; panels: readonly PackPanelDeclaration[] }
export interface PackNavigationMapperOptions { resolveLabel?: (labelKey: string) => string; state?: ShellNavigationState; roleVocabulary: RoleVocabulary; heldRoles: HeldRoleSet }

const roleReference = (name: string): RoleReference | undefined => {
  const separator = name.lastIndexOf('/')
  if (separator <= 0 || separator === name.length - 1) return undefined
  const vocabulary = name.slice(0, separator)
  if (vocabulary !== 'sys.platform-roles' && vocabulary !== 'tax.roles') return undefined
  return { vocabulary, name: name.slice(separator + 1) }
}
function permitted(entry: PackNavigationAction, vocabulary: RoleVocabulary, held: HeldRoleSet) {
  if (!entry.permittedRoles.length) return true
  const roles = entry.permittedRoles.map(roleReference)
  return roles.every((role): role is RoleReference => role !== undefined) && roleGateAllows({ requiredRoles: roles }, vocabulary, held)
}

/** The only React projection mapper from the api#58 declaration into shell render state. */
export function mapPackNavigationDeclaration(declaration: PackNavigationDeclaration, options: PackNavigationMapperOptions): ShellNavigationViewModel {
  const label = options.resolveLabel ?? (key => key)
  const state = options.state ?? {}
  const action = (entry: PackNavigationAction): ShellActionViewModel => ({ id: entry.id, label: label(entry.verbKey), icon: entry.icon, binding: entry.binding, shortcut: entry.shortcut })
  return {
    workspaces: declaration.seedWorkspaces.map(workspace => {
      const declaredActions = workspace.createActions ?? []
      const allowedActions = declaredActions.filter(entry => permitted(entry, options.roleVocabulary, options.heldRoles))
      const deniedGuidance = declaredActions.filter(entry => !permitted(entry, options.roleVocabulary, options.heldRoles)).map(entry => state.capabilityGuidanceByBinding?.[entry.binding]).find(Boolean)
      return { id: workspace.id, label: label(workspace.labelKey), icon: workspace.icon, count: workspace.countQueryRef ? state.counts?.[workspace.countQueryRef] : undefined,
        groups: (workspace.groups ?? []).map(group => ({ id: group.id, label: label(group.labelKey), items: group.itemIds.map(id => state.items?.[id] ?? { id, label: id }), addAction: group.addAction && permitted(group.addAction, options.roleVocabulary, options.heldRoles) ? action(group.addAction) : undefined })),
        createActions: allowedActions.map(action), defaultCreateActionId: state.defaultCreateActionByWorkspace?.[workspace.id], createGuidance: deniedGuidance, documentSpine: workspace.documentSpine ?? [], recent: state.recentByWorkspace?.[workspace.id] ?? [], suggested: state.suggestedByWorkspace?.[workspace.id] }
    }),
    modes: (declaration.modeSwitch?.modes ?? []).map(mode => ({ id: mode.id, label: label(mode.labelKey), workspaceIds: mode.workspaceIds })),
    panels: declaration.panelSet ?? [],
  }
}

File: projections/blazor/ui/hlp.ui.app-shell/PackNavigationMapper.cs
using Harborline.Contracts.Authorization;

namespace Harborline.UIAdapters.Blazor.Components.Layout;

public static class PackNavigationMapper
{
    /// <summary>The only Blazor projection mapper from the api#58 declaration into shell render state.</summary>
    public static ShellNavigationViewModel Map(PackNavigationDeclaration declaration, Func<string, string>? resolveLabel, ShellNavigationState? state, RoleVocabulary vocabulary, HeldRoleSet heldRoles)
    {
        var label = resolveLabel ?? (key => key);
        state ??= new();
        ShellActionViewModel Action(PackNavigationAction entry) => new(entry.Id, label(entry.VerbKey), entry.Icon, entry.Binding, entry.Shortcut);
        bool Permitted(PackNavigationAction entry)
        {
            if (entry.PermittedRoles.Count == 0) return true;
            var roles = entry.PermittedRoles.Select(ParseRole).ToArray();
            return roles.All(role => role is not null) && RoleGateResolver.Allows(new(roles!), vocabulary, heldRoles);
        }
        var workspaces = declaration.SeedWorkspaces.Select(workspace =>
        {
            var declared = workspace.CreateActions ?? [];
            var allowed = declared.Where(Permitted).Select(Action).ToArray();
            var guidance = declared.Where(entry => !Permitted(entry)).Select(entry => state.CapabilityGuidanceByBinding?.GetValueOrDefault(entry.Binding)).FirstOrDefault(value => value is not null);
            return new ShellWorkspaceViewModel(workspace.Id, label(workspace.LabelKey), (workspace.Groups ?? []).Select(group => new ShellGroupViewModel(group.Id, label(group.LabelKey), group.ItemIds.Select(id => state.Items?.GetValueOrDefault(id) ?? new ShellNavItem(id, id)).ToArray(), group.AddAction is not null && Permitted(group.AddAction) ? Action(group.AddAction) : null)).ToArray(), workspace.CountQueryRef is null ? null : state.Counts?.GetValueOrDefault(workspace.CountQueryRef), allowed, state.DefaultCreateActionByWorkspace?.GetValueOrDefault(workspace.Id), guidance, workspace.DocumentSpine ?? [], state.RecentByWorkspace?.GetValueOrDefault(workspace.Id) ?? [], state.SuggestedByWorkspace?.GetValueOrDefault(workspace.Id));
        }).ToArray();
        var modes = (declaration.ModeSwitch?.Modes ?? []).Select(mode => (mode.Id, label(mode.LabelKey), mode.WorkspaceIds)).ToArray();
        return new(workspaces, modes, declaration.PanelSet ?? []);
    }

    private static RoleReference? ParseRole(string value)
    {
        var separator = value.LastIndexOf('/');
        if (separator <= 0 || separator == value.Length - 1) return null;
        var vocabulary = value[..separator];
        return vocabulary is RoleVocabularies.Platform or RoleVocabularies.Domain ? new(vocabulary, value[(separator + 1)..]) : null;
    }
}
